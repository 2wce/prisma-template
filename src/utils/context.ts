import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()
export interface Context {
  prisma: PrismaClient
  request: any
}

export function createContext(request: any) {
  return {
    ...request,
    prisma,
  }
}
