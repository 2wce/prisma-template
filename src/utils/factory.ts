import { Prisma, PrismaClient } from '@prisma/client'
import * as Factory from 'factory.ts'
import { commerce, internet, name } from 'faker'

const prisma = new PrismaClient()

export interface TableList {
  TABLE_NAME: string
}

export const userFactory = Factory.Sync.makeFactory<Prisma.UserCreateManyInput>(
  {
    name: Factory.Sync.each(() => name.firstName()),
    surname: Factory.Sync.each(() => name.lastName()),
    email: Factory.Sync.each((seq) => `seq_${seq}_${internet.email()}`),
    password: Factory.Sync.each(() => internet.password()),
  },
)

export const postFactory = Factory.Sync.makeFactory<Prisma.PostCreateManyInput>(
  {
    title: Factory.Sync.each(() => commerce.productName()),
    content: Factory.Sync.each(() => commerce.productDescription()),
    published: true,
  },
)

export const clearData = async () => {
  try {
    const posts = await prisma.post.deleteMany()
    const users = await prisma.user.deleteMany()

    console.log({ posts, users })
  } catch (error) {
    console.error(error)
    return error
  }
}
