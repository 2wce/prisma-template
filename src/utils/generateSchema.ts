import type { Dirent } from "node:fs";
import { readdir } from "node:fs/promises";
import path from "path";

const isSchema = (file: Dirent) => file.name.includes("schema");

// Define the path to the directory that contains the resolvers.
const directoryPath = path.join(import.meta.dir, "../app");

// Read all files in the directory.
// This is done recursively, so it includes files in subdirectories.
// The 'withFileTypes' option is set to 'true', so the returned array includes 'Dirent' objects instead of strings.
const files = await readdir(directoryPath, {
	recursive: true,
	withFileTypes: true,
});

// Filter the files to get only the schema.
const schemaFiles = files.filter((file) => isSchema(file));

// Define an array to hold the schemas.
const schemas = [];

for (const schemaFile of schemaFiles) {
	const { default: schemaExport } = await import(
		`${directoryPath}/${schemaFile.name}`
	);

	const schema = await Bun.file(schemaExport, { type: "graphql" }).text();

	schemas.push(schema);
}

// Return the merged schema.
const schema = schemas.join("");
const spacing = "\n\n  ";

// Match the contents of the Mutation types
const mutationMatches = schema.match(/type Mutation \{([^}]*)\}/gs);

// Extract the fields from the Mutation types and trim them
const mutationFields = mutationMatches?.map((match) =>
	match.replace(/type Mutation \{|\}/g, "").trim(),
);

// Match the contents of the Query types
const queryMatches = schema.match(/type Query \{([^}]*)\}/gs);

// Extract the fields from the Query types and trim them
const queryFields = queryMatches?.map((match) =>
	match.replace(/type Query \{|\}/g, "").trim(),
);

// Trim the fields to remove leading and trailing whitespace
const mutation = mutationFields?.join(spacing).trim();
const query = queryFields?.join(spacing).trim();
const rest = schema
	.replace(/type Mutation \{([^}]*)\}/gs, "")
	.replace(/type Query \{([^}]*)\}/gs, "")
	.trim();

// Combine the Mutation and Query types
const combinedSchema = `
type Mutation {
  ${mutation}
}

type Query {
  ${query}
}

${rest}
`.trim();

const outputPath = path.join(import.meta.dir, "../schema.graphql");

await Bun.write(outputPath, combinedSchema);
