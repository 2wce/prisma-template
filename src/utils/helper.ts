import { ResetPasswordInput } from '../generated'

export const formatError = (nameOfFunction: string, error: Error) => {
  return process.env.NODE_ENV === 'test'
    ? ''
    : console.log(`----${nameOfFunction}----`, error)
}

// RegExp to check if email is valid
export const emailRegExp =
  /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

export const generateRandomOtp = (noOfDigits: number): number => {
  const multiplier = Math.pow(10, noOfDigits - 1)
  return Math.floor(multiplier + Math.random() * (9 * multiplier))
}

export const hasValidResetPasswordInput = (
  params: ResetPasswordInput,
): boolean => {
  const { password, passwordConfirmation, code } = params

  // check if all values are present
  const isAllValuesPresent = Boolean(
    password &&
      passwordConfirmation &&
      code &&
      password.length > 0 &&
      passwordConfirmation.length > 0,
  )

  // check if password and password confirmation are equal
  const isPasswordEqual = Boolean(password === passwordConfirmation)

  return isAllValuesPresent && isPasswordEqual
}
