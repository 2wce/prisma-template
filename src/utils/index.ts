export * from './jwt'
export * from './prisma'
