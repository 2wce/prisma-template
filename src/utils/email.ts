import { createTransport } from 'nodemailer'
import { formatError } from './helper'

export const sendEmail = async (
  recipientEmail: string,
  html: string,
  subject: string,
) => {
  try {
    // create reusable transporter object using the default SMTP transport
    const transporter = createTransport({
      host: process.env.SMTP_HOST,
      port: process.env.SMTP_PORT,
      secure: true, // true for 465, false for other ports
      auth: {
        user: process.env.SMTP_USERNAME, // generated ethereal user
        pass: process.env.SMTP_PASSWORD, // generated ethereal password
      },
    })

    // send mail with defined transport object
    const info = await transporter.sendMail({
      from: `Admin <${process.env.NO_REPLY_EMAIL}>`, // sender address
      to: recipientEmail, // list of receivers
      subject, // Subject line
      html, // html body
    })

    console.log('Message sent:', info)

    return info
  } catch (error) {
    formatError('sendEmail', error)
    return error
  }
}
