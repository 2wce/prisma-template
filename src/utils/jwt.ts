import { compare, hash } from 'bcryptjs'
import { sign, verify } from 'jsonwebtoken'
import { clone } from 'lodash'
import { Context } from './prisma'

interface Token {
  id: string
}

type User = {
  id: number
}

// get user id from auth token
export function getUserId({ request }: Context): string | undefined {
  if (request && request.headers) {
    const Authorization =
      request.headers.authorization || request.headers.Authorization

    if (Authorization) {
      const token = Authorization.replace('Bearer ', '')

      const verifiedToken = verify(
        token,
        process.env.JWT_SECRET as string,
      ) as Token

      return verifiedToken && verifiedToken.id
    }
  }
  return undefined
}

// issue new token based on payload
export const issue = (payload: string | User | Buffer, jwtOptions = {}) => {
  return sign(clone(payload), process.env.JWT_SECRET as string, jwtOptions)
}

export const isHashed = (password: string) => {
  if (typeof password !== 'string' || !password) {
    return false
  }

  return password.split('$').length === 4
}

export const hashPassword = (password: string): Promise<string | null> => {
  return new Promise((resolve, reject) => {
    if (!password || isHashed(password)) {
      resolve(null)
    } else {
      hash(`${password}`, process.env.SALT as string, (err, hash) => {
        if (err) {
          return reject(err)
        }
        resolve(hash)
      })
    }
  })
}

export const validatePassword = (password: string, hash: string) => {
  return compare(password, hash)
}
