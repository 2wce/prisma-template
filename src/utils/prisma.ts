import { PrismaClient } from "@prisma/client";
import http from "http";
import type { MockProxy } from "jest-mock-extended";
import { mockDeep } from "jest-mock-extended";
import { getUserId } from "./jwt";

export const prisma = new PrismaClient();

export interface ContextEvent {
	headers: {
		authorization: string;
		Authorization: string;
	};
}

export interface ContextArgs {
	req: http.IncomingMessage;
	res: http.ServerResponse;
}
export interface Context {
	prisma: PrismaClient;
	req: http.IncomingMessage;
	res: http.ServerResponse;
	userId?: string;
}

export type MockContext = {
	prisma: MockProxy<PrismaClient>;
};

export const createMockContext = (): MockContext => {
	return {
		prisma: mockDeep<PrismaClient>(),
	};
};

// add prisma to context for resolvers
export function createContext({ req, res }: ContextArgs) {
	return {
		req,
		res,
		userId: getUserId(req),
		prisma,
	};
}
