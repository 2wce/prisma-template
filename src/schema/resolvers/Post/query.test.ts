import { getPrismaTestInstance } from '../../../utils'
import postResolvers from './index'

describe('Successful post queries', () => {
  it('can get filtered posts with valid search string', async () => {
    // setup
    const {
      Query: { filterPosts },
    } = postResolvers
    const args = { searchString: 'Subscribe to' }
    const context = {
      prisma: getPrismaTestInstance,
      request: {},
    }

    // test
    const posts = await filterPosts({}, args, context)

    // assert
    expect(posts).toMatchSnapshot()
  })

  it('can get feed with published posts', async () => {
    // setup
    const {
      Query: { feed },
    } = postResolvers
    const context = {
      prisma: getPrismaTestInstance,
      request: {},
    }

    // test
    const posts = await feed({}, {}, context)

    // assert
    expect(posts).toMatchSnapshot()
  })

  it('can get post with valid id', async () => {
    // setup
    const {
      Query: { post },
    } = postResolvers
    const context = {
      prisma: getPrismaTestInstance,
      request: {},
    }
    const args = { where: { id: 2 } }

    // test
    const result = await post({}, args, context)

    // assert
    expect(result).toMatchSnapshot()
  })
})

afterAll(async done => {
  await getPrismaTestInstance.disconnect()
  done()
})
