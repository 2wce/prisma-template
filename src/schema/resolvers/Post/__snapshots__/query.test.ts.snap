// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Successful post queries can get feed with published posts 1`] = `
Array [
  Object {
    "authorId": 1,
    "content": "https://graphqlweekly.com/",
    "id": 2,
    "published": true,
    "title": "Subscribe to GraphQL Weekly for community news ",
  },
]
`;

exports[`Successful post queries can get filtered posts with valid search string 1`] = `
Array [
  Object {
    "authorId": 1,
    "content": "https://graphqlweekly.com/",
    "id": 2,
    "published": true,
    "title": "Subscribe to GraphQL Weekly for community news ",
  },
]
`;

exports[`Successful post queries can get post with valid id 1`] = `
Object {
  "authorId": 1,
  "content": "https://graphqlweekly.com/",
  "id": 2,
  "published": true,
  "title": "Subscribe to GraphQL Weekly for community news ",
}
`;
